src/shims/x86/gfni.rs

```diff
@@ -0,0 +1,215 @@
+    right: &OpTy<'tcx>,
+    imm8: &OpTy<'tcx>,
+    dest: &MPlaceTy<'tcx>,
+    inverse: bool,
+) -> InterpResult<'tcx, ()> {
+    let (left, left_len) = this.project_to_simd(left)?;
+    let (right, right_len) = this.project_to_simd(right)?;
+    let (dest, dest_len) = this.project_to_simd(dest)?;
+
+    assert_eq!(dest_len, right_len);
+    assert_eq!(dest_len, left_len);
+
+    let imm8 = this.read_scalar(imm8)?.to_u8()?;
+
+    // Each 8x8 bit matrix gets multiplied with eight bit vectors.
+    // Therefore, the iteration is done in chunks of eight.
+    for i in (0..dest_len).step_by(8) {
+        // Get the bit matrix.
+        let mut matrix = [0u8; 8];
+        for j in 0..8 {
+            matrix[usize::try_from(j).unwrap()] =
+                this.read_scalar(&this.project_index(&right, i.wrapping_add(j))?)?.to_u8()?;
+        }
+
+        // Multiply the matrix with the vector and perform the addition.
+        for j in 0..8 {
+            let index = i.wrapping_add(j);
+            let left = this.read_scalar(&this.project_index(&left, index)?)?.to_u8()?;
+            let left = if inverse { TABLE[usize::from(left)] } else { left };
+
+            let mut res = 0;
+
+            // Do the matrix multiplication.
+            for bit in 0u8..8 {
+                let mut b = matrix[usize::from(bit)] & left;
+
+                // Calculate the parity bit.
+                b = (b & 0b1111) ^ (b >> 4);
+                b = (b & 0b11) ^ (b >> 2);
+                b = (b & 0b1) ^ (b >> 1);
+
+                res |= b << 7u8.wrapping_sub(bit);
+            }
+
+            // Perform the addition.
+            res ^= imm8;
+
+            let dest = this.project_index(&dest, index)?;
+            this.write_scalar(Scalar::from_u8(res), &dest)?;
+        }
+    }
+
+    Ok(())
+}
+
+/// A lookup table for computing the inverse byte for the inverse affine transformation.
+/// It was generated using the following program on a GFNI-enabled computer:
+/// ```rust
+/// #![feature(
+///     stdarch_x86_avx512,
+///     avx512_target_feature,
+/// )]
+/// 
+/// use std::arch::x86_64::*;
+/// 
+/// 
+/// #[target_feature(enable = "gfni,avx2")]
+/// unsafe fn get_value(x: u8) -> u8 {
+///     let ident = _mm_set1_epi64x(i64::from_le_bytes([
+///         0b1000_0000,
+///         0b0100_0000,
+///         0b0010_0000,
+///         0b0001_0000,
+///         0b0000_1000,
+///         0b0000_0100,
+///         0b0000_0010,
+///         0b0000_0001,
+///     ]));
+///     let vector = _mm_set1_epi8(x as i8);
+///     let r = _mm_gf2p8affineinv_epi64_epi8::<0>(vector, ident);
+///     let r = _mm_extract_epi64::<0>(r) & 0xff;
+///     r as u8
+/// }
+/// 
+/// pub fn main() {
+///     for i in 0..=u8::MAX {
+///         unsafe {
+///             print!("0x{:02x}, ", get_value(i));
+///         }
+///         if i & 15 == 15 {
+///             println!();
+///         }
+///     }
+/// }
+/// ```
+#[rustfmt::skip]
+static TABLE: [u8; 256] = [
```

Comment:
I am not happy with just carrying a random bunch of blobs here. Does the official Intel spec for these instruction refer to this table somehow? Or does it describe these instructions algorithmically? If it is the latter, can we use the algorithmic version in Miri?

Reply:
The intrinsics guide only specifies that the byte is inverted, even inside the pseudocode.
I did find a source which includes a definition of the inverse byte, and created the table based on that.

Reply:
Please give the source.

but if possible I'd like to not have the table at all.

Reply:
The link is mentioned in `gf2p8_mul`. Should I add a reference here aswell?

Reply:
I see you made it compute the table in the static initializer now, great. :)

Yes the link should be next to the code that was written based on that source.

Reply:
Done.